transport.go

```diff
@@ -202,6 +219,37 @@ func (t *transport) Dial(ctx context.Context, raddr ma.Multiaddr, p peer.ID) (tp
 	return conn, nil
 }
 
+func (t *transport) holePunch(network string, addr *net.UDPAddr) error {
+	pconn, err := t.connManager.Dial(network, addr)
+	if err != nil {
+		return err
+	}
+	defer pconn.DecreaseCount()
+
+	conn := pconn.UDPConn
+
+	payload := make([]byte, 64)
+	_, err = rand.Read(payload)
+	if err != nil {
+		return err
+	}
+
+	_, err = conn.WriteToUDP(payload, addr)
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < 10; i++ {
+		time.Sleep(time.Duration(rand.Intn(i * int(10*time.Millisecond))))
+		_, err := conn.WriteToUDP(payload, addr)
```

Comment:
Better: generate a new random payload for each packet. Some middleboxes might inspect the packet (e.g. to detect the protocol being used) and decide to drop it if they don't like it for any reason.

Reply:
Very good point, will do.

Reply:
I think we should also make sure to not send too many packets here. Sending a packet more often than every 10ms seems unnecessary. Maybe make it `10ms + rand.Intn(10ms)`?
That would also make sure that this function will run for (on average) 150ms instead of just 50ms.

Reply:
Sure, let's not flood the network with packets.